### PR TITLE
Added support for declared databases that become services

### PR DESCRIPTION
**Pull request by davidthor**
## Overview

Added support for declarative databases that will transform into services for a simpler developer experience

## Changes

Updated the component spec to allow for declared databases and updated transformer logic to create service nodes with the `-db` suffix. Also updated interpolation logic to populate services/tasks based on new expression syntax `${{ databases.<name>.connection_string }}`.
1. Update component spec to allow for declared databases.
2. Updated transformer logic to convert databases into a service node with the `-db` suffix.
3. Add interpolation logic to allow new expressions such as `${{ databases.<name>.connection_string }}`.
4. Added a new resource node type called databases. This is not used heavily now, but lays the ground work for more complicated databases when we need to no longer just treat them as services.
5. Connection strings can bet set to secrets to allow them to be overridden at deploy time from the CLI.
6. Connection string must be a URI.

## Tests

1. Added new test cases to both the superset and spec tests.
2. Tested a bunch of variations on the nodejs sample app like so
```
name: nodejs
description: A simple todo application built with Node.js, Express, a postgres database, and a static html page
homepage: https://github.com/architect-templates/nodejs
keywords:
  - architect
  - examples
  - postgres
  - express
  - nodejs

# Add secrets to be used by different services. For more information:
# https://docs.architect.io/deployments/secrets/
secrets:
  db_user:
    description: Root user to assign to the component's database
    default: architect
  db_pass:
    description: Root password to assign to the component's database
    default: secret
  db_name:
    description: Name of the database the component will store content in
    default: api-db
  db_port:
    description: Port for the database
    default: 5432
  api_port:
    description: Port for api
    default: 8080

databases:
  main:
    type: postgres:12

databases:
  main2:
    type: postgres:12
    connection_string: https://test.com

services:
  ### This service is the actual Node.js component ###
  api:
    # Build context is the location of your application relative to the architect.yml file
    build:
      context: .
    interfaces:
      main:
        port: ${{ secrets.api_port }}
        ingress:
          subdomain: api
    # The liveness_probe makes sure the application is healthy - if not, it will re-start the application
    liveness_probe:
      command: curl --fail 0.0.0.0:8080
      interval: 5s
      failure_threshold: 3
    # Set environment variables for your application here
    environment:
      DB_ADDR: ${{ databases.main.dsn }}
      USERNAME: ${{ databases.main.host }}
      PORT: ${{ secrets.api_port }}
    # Debug block is used to enable hot-reloading to make real-time changes to an
    #   application while running `architect dev`. For more info, see -
    #   https://docs.architect.io/components/local-configuration/#hot-reloading
    debug:
      command: npm run dev
      build:
        args:
          NODE_ENV: development
      volumes:
        src:
          # The `WORKDIR` defined in the Dockerfile (/usr/src/app)
          mount_path: /workspace
          # Path to source code relative to the architect.yml file
          host_path: .
```

### CODE REVIEW

File: src/dependency-manager/spec/transform/component-transform.ts [hunk 1]
```diff
@@ -77,7 +75,10 @@ export const transformComponentSpec = (spec: ComponentSpec): ComponentConfig =>
     secrets,
     outputs,
 
-    services,
+    services: {
+      ...services,
+      ...databases,
```
Comment (tjhiggins) [inline src/dependency-manager/spec/transform/component-transform.ts:80]: I guess this will be decided by the database node conv

**Review by tjhiggins — CHANGES_REQUESTED** [review]

File: src/dependency-manager/manager.ts [hunk 2]
```diff
@@ -122,6 +122,24 @@ export default abstract class DependencyManager {
...
+
+      const database_regex = new RegExp(`\\\${{\\s*(dependencies\\.(?<dependency_name>${ComponentSlugUtils.RegexBase})\\.)?databases\\.(?<database_name>${Slugs.ArchitectSlugRegexBase})\\.(?<database_key>${Slugs.ArchitectSlugRegexBase})`, 'g');
+      while ((matches = database_regex.exec(resource_string)) !== null) {
+        if (!matches.groups) continue;
+
+        const { dependency_name, database_name } = matches.groups;
+
+        const dependency = component_configs.find(c => c.name === dependency_name) || component;
+
+        const to = buildNodeRef(dependency, 'services', `${database_name}-db`);
```
[the remark this answers is not in the document]
Comment (Zach-at-Arch) [reply src/dependency-manager/manager.ts:134]: This is going to determine how we handle [ResourceTypes](https://github.com/architect-team/architect-cli/blob/91441202a9177e4df32f94723419d6845ef16ffd/src/dependency-manager/spec/utils/slugs.ts#L142), and ultimately how we will handle other entities going forward. 

I like the idea of adding databases to `ResourceType` since those are one of the more important entities that people are concerned about in their applications. This also opens us up to other resources - I can see managed cloud resources as `cloud` ResourceType. 

Also, I think we should drop the `-db` from the name and just keep it as ${database_name}

File: src/dependency-manager/spec/database-spec.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  })
+  description?: string;
+
+  @IsString()
+  @JSONSchema({
+    ...ExpressionOr({ type: 'string', pattern: Slugs.ComponentDatabaseValidator.source }),
+    description: 'The type engine and version of database software needed for data storage.',
+    errorMessage: Slugs.ComponentDatabaseDescription,
+  })
+  type!: string;
```
Comment (Zach-at-Arch) [inline src/dependency-manager/spec/database-spec.ts:29]: We won't need the engine and version here - I think this should be `engine` and not `type` - `engine` will be more specific and prompt users to enter something like "postgres" or "mysql", while `type` might prompt a user to instead give "SQL" or "NoSQL".

File: src/dependency-manager/spec/database-spec.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import { AnyOf, ExpressionOr } from './utils/json-schema-annotations';
+import { Slugs } from './utils/slugs';
+
+export class DatabaseSpec {
+  @IsOptional()
+  @JSONSchema({
+    type: 'string',
+    pattern: Slugs.ArchitectSlugValidator.source,
+    errorMessage: Slugs.ArchitectSlugDescription,
+    description: 'A specific service name which will override the database name specified in the component.',
```
Comment (Zach-at-Arch) [inline src/dependency-manager/spec/database-spec.ts:12]: why would we need this? This seems extremely confusing for a new user on the first touch. I'm assuming it's for host overrides.

I don't know why a user would want to override the database name they specify in the component - I assume they would want to keep it named the same so they know what it is

File: src/dependency-manager/spec/database-spec.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    errorMessage: Slugs.ComponentDatabaseDescription,
+  })
+  type!: string;
+
+  @IsOptional()
+  @JSONSchema({
+    ...ExpressionOr(AnyOf('null', 'string')),
+    description: 'The connection string of an existing database to use instead of provisioning a new one',
+  })
+  dsn?: null | string;
```
[the remark this answers is not in the document]
Comment (Zach-at-Arch) [reply src/dependency-manager/spec/database-spec.ts:36]: On this note - I think `url` might be better here instead of `dsn`

File: src/dependency-manager/spec/database-spec.ts (hunk 3 above)
Comment (Zach-at-Arch) [reply src/dependency-manager/spec/database-spec.ts:29]: Or, to stay more consistent with DatabaseContext, this could be `protocol`

File: src/dependency-manager/spec/transform/database-transform.ts [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+};
+
+const SupportedDatabases: SupportedDatabaseType[] = [
+  {
+    engine: 'postgres',
+    versions: {
+      min: 10,
+    },
+    spec: {
+      image: 'postgres:15',
```
[the remark this answers is not in the document]
Comment (Zach-at-Arch) [reply src/dependency-manager/spec/transform/database-transform.ts:26]: Unless we go off the assumption all minor versions are supported within a given major version

File: src/dependency-manager/manager.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidthor) [reply src/dependency-manager/manager.ts:134]: This was just an oversight on my part. I think what TJ wrote is right for the node ref.

File: src/dependency-manager/manager.ts [hunk 7]
```diff
@@ -319,13 +338,25 @@ export default abstract class DependencyManager {
...
       }
+
       const service_ref = buildNodeRef(component_config, 'services', service_name);
       for (const [interface_name, interface_config] of Object.entries(service.interfaces)) {
         const interface_ref = `services.${service_name}.interfaces.${interface_name}`;
 
         const architect_host = service_ref;
         const architect_port = `${interface_ref}.external_port`;
 
+        // TODO: Remove once databases get their own node type
```
[the remark this answers is not in the document]
Comment (davidthor) [reply src/dependency-manager/manager.ts:349]: It would only really be necessary when we eventually add provisioning of managed DBs (e.g. RDS) so that we know to fulfill the node differently. Since this is just transforming it into a service its not necessary.

File: src/dependency-manager/spec/database-spec.ts (hunk 4 above)
Comment (davidthor) [reply src/dependency-manager/spec/database-spec.ts:12]: I don't remember the use case exactly, but @tjhiggins can chime in. He was the one who added that to services so I just tried to mirror it for now.

File: src/dependency-manager/spec/database-spec.ts (hunk 3 above)
Comment (davidthor) [reply src/dependency-manager/spec/database-spec.ts:29]: I avoided that because I consider the the value to be `<engine>:<version>`. If we pull version into a different field then I think engine makes sense, but I was considering `type` to be the joint value. This also helps us mirror docker tag conventions which aids us in generating dockerized DB instances.

File: src/dependency-manager/spec/database-spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (davidthor) [reply src/dependency-manager/spec/database-spec.ts:36]: I think most of the time DSN is referenced as this format:

```
<protocol>://<user>:<pass>@<host>:<port>/<database>
```

while URLs are usually expected to be this:

```
<protocol>://<host>:<port>
```

I know that's not exactly what `url` means for service interfaces today since we also optionally inject the username and password, but I thought the added path made it feel different. That being said, we could definitely have both terms available for both resource types (services and DBs) and just favor different language in our docs when talking about them based on what developers are expecting or searching for naturally.

File: src/dependency-manager/spec/transform/database-transform.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (davidthor) [reply src/dependency-manager/spec/transform/database-transform.ts:26]: This `image` value is actually ignored for the most part and overwritten here:
https://github.com/architect-team/architect-cli/pull/790/files#diff-8104f1db58060b55c205e2361e249be108e45c0cfebcbee720451df972544da2R105

**Review by davidthor — COMMENTED** [review]
Handing this over to @Zach-at-Arch now. Just wanted to implement the spec changes to get the ball rolling while we decide on UI changes.

File: src/dependency-manager/spec/transform/database-transform.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (tjhiggins) [reply src/dependency-manager/spec/transform/database-transform.ts:26]: > Unless we go off the assumption all minor versions are supported within a given major version

I think this is a fine assumption. Unlikely that env variables/ports change between minor versions.

To confirm the following 3 databases would use the same service definition with the only diff being the image:

```yaml
databases:
  postgres-9-4:
    engine: postgres:9.4
  postgres-9-6:
    engine: postgres:9.6
  postgres-9:
    engine: postgres:9
```

[the remark this answers is not in the document]
Comment (davidthor) [reply src/dependency-manager/spec/transform/database-transform.ts:26]: That’s what I was assuming. There will likely be a few edge cases but I think that works for the most part.

File: src/dependency-manager/spec/database-spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (mueschm) [reply src/dependency-manager/spec/database-spec.ts:36]: Thinking about this more I think we should change it to `connectionString`. DSN is generally used in places where you have the configuration stored somewhere else and are just referencing the configuration by name. While it can be used as the connection string, in my experience dsn is not as well understood anymore.

The other thing is that there are generally two types of connection strings. The first is your standard url but then other systems like .NET actually use ODBC which uses its own connection string syntax. To make sure we support as many formats as possible we should also add other values that can be used in interpolation.

So a user may configure a `connectionString` as part of their override, but they should be able to access it for ODBC connections via `connectionStringODBC`. The same thing should be allowed in the reverse. Otherwise, you may need to declare two databases each with its own connection string in order to support every use case.

File: src/dependency-manager/spec/database-spec.ts (hunk 3 above)
Comment (mueschm) [reply src/dependency-manager/spec/database-spec.ts:29]: I am not sold on type here either because I don't feel it does a good enough job of conveying exactly what the value is on first glance. Though I can't think of something better than type at least right now.

File: src/dependency-manager/spec/database-spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (davidthor) [reply src/dependency-manager/spec/database-spec.ts:36]: `connectionString` is a bit wordy, but its the only other keyword that I've thought makes sense. We can make that the input value if you want.

For outputs (e.g. referencing `${{ databases.name.connectionString }}`) I think we should support both names since that allows someone to be brief or verbose. Should be easy enough to implement too.

File: src/dependency-manager/spec/database-spec.ts (hunk 3 above)
Comment (davidthor) [reply src/dependency-manager/spec/database-spec.ts:29]: I don't think we need to overthink it. The configuration for a database is so minimal that it'll be easy no matter what the keyword is.

**Comment by mueschm** [comment]
Alright, everything is functional, and most comments have been addressed. The main things left are all-around user experience.

1. Should we rename dsn to connection_string?
2. In an earlier conversation, we talked about the fact services use url while databases use dsn. Do we need them to match or are we okay with the difference since databases do not have url?

Personally, I am leaning towards just using `connection_string` and allowing services and databases to have different values. Though curious what others think.

**Comment by davidthor** [comment]
> Personally, I am leaning towards just using connection_string and allowing services and databases to have different values.
For the odbc problem, I think we need to support them in order to make it easy to integrate with Java and .NET which both use this format. I also want to make it so that if you set connection_string or connection_string_odbc that we are able to convert from one format to the other.
Though curious what others think.

I like your conclusions here. `connection_string` is fine by me. Perhaps URL should be allowed as well, but could means the URL without the credentials so that something like this is possible:

```
environment:
  DB_URL: ${{ databases.main.url }}
  DB_USER: ${{ databases.main.username }}
  DB_PASS: ${{ databases.main.password }}

---
# or
environment:
  DB_ADDR: ${{ databases.main.connection_string }}
```

File: src/dependency-manager/config/component-context.ts [hunk 8]
```diff
@@ -4,6 +4,16 @@ import { ServiceInterfaceConfig } from './service-config';
...
 export type OutputValue = string | number | boolean | null;
 
+export interface DatabaseContext {
+  protocol: string;
+  host: string;
+  port: number | string;
+  username: string;
+  password: string;
+  database: string;
+  connection_string: string;
```
[the remark this answers is not in the document]
Comment (davidthor) [reply src/dependency-manager/config/component-context.ts:14]: @mueschm and I discussed this at length, and connection string is the most widely accepted term for that value in the context of databases and ORMs so we decided to stick with it. It's better for it to align with expectations than to arbitrarily conform to a vague word like `url`.

[the remark this answers is not in the document]
Comment (tjhiggins) [reply src/dependency-manager/config/component-context.ts:14]: Can we support both? Seems pretty confusing to use a different value for interpolation compared to services.

[the remark this answers is not in the document]
Comment (davidthor) [reply src/dependency-manager/config/component-context.ts:14]: I disagree, but I'm never opposed to aliases like that. I'm happy to support both.

[the remark this answers is not in the document]
Comment (tjhiggins) [reply src/dependency-manager/config/component-context.ts:14]: Lets do both. Typeorm and Sequelize use `url` in the config and I find that more intuitive. Seems like that is because postgres/mysql npm pkgs also use url.

File: src/dependency-manager/spec/transform/component-transform.ts [hunk 9]
```diff
@@ -55,6 +56,12 @@ export const transformComponentSpec = (spec: ComponentSpec): ComponentConfig =>
   const services = transformDictionary(transformServiceSpec, spec.services, spec.metadata);
   const tasks = transformDictionary(transformTaskSpec, spec.tasks, spec.metadata);
   const dependencies = spec.dependencies || {};
+  const service_databases = transformDictionary(transformDatabaseSpecToServiceSpec, spec.databases, spec.metadata);
+  const databases = transformDictionary(transformDatabaseSpec, spec.databases, spec.metadata);
+  for (const [key, value] of Object.entries(service_databases)) {
+    service_databases[`${key}-db`] = value;
```
[the remark this answers is not in the document]
Comment (mueschm) [reply src/dependency-manager/spec/transform/component-transform.ts:62]: The problem is that `--` is not valid as a name in our system. So after we generate the service our validation throws an error. This should only be a problem for a bit until we add in support for fully managed databases and then this will go away since we will have code that does not treat databases as services.

File: src/dependency-manager/config/component-context.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (mueschm) [reply src/dependency-manager/config/component-context.ts:14]: So I added support for url via interpolation, but connection_string is the only valid option for the spec. I think it would be weird to have both there. Unless you guys disagree.

[the remark this answers is not in the document]
Comment (tjhiggins) [reply src/dependency-manager/config/component-context.ts:14]: That's fine. Host overrides for services should probably use the same mechanism, but we can cross that bridge later.

File: src/dependency-manager/spec/utils/slugs.ts [hunk 10]
```diff
@@ -20,6 +20,8 @@ export class Slugs {
   public static INSTANCE_DELIMITER = '@';
   public static SLUG_CHAR_LIMIT = 32;
 
+  public static DB_SUFFIX = '-managed-db';
```
Comment (tjhiggins) [inline src/dependency-manager/spec/utils/slugs.ts:23]: I'm not sold on this name since we don't manage the db in the preview case. Would prefer `-db`, but I still don't love that someone could create a conflict via:

```yaml
databases:
  mysql:

services:
  mysql-db:
```

Obviously unlikely that someone names the services in this way. If we are going to move forward with this services block merging then we should at least error if a conflict is generated.

File: src/dependency-manager/utils/parser.ts [hunk 11]
```diff
@@ -197,6 +196,12 @@ export class ArchitectParser {
             value = node.arguments[0].value.trim();
           } else if (node.callee.value === 'startsWith') {
             value = node.arguments[0].value.startsWith(node.arguments[1].value);
+          } else if (node.callee.value === 'parseUrl') {
+            try {
+              value = (new URL(node.arguments[0].value) as any)[node.arguments[1].value as string];
+            } catch {
+              value = '';
```
Comment (tjhiggins) [inline src/dependency-manager/utils/parser.ts:203]: Should this silently error?

**Review by tjhiggins — CHANGES_REQUESTED** [review]
